### PR TITLE
Python: properly support keyword only arguments

### PR DESCRIPTION
**Pull request by RasmusWL**
## Main purpose of this PR:

To properly support keyword-only parameters. That is done in https://github.com/github/codeql/pull/3362/commits/c508e89a0059a0d3b352a3e7d944ab588bb8b78a.

## Secondary purpose: 

Properly support default values for keyword-only parameters. That is done in https://github.com/github/codeql/pull/3362/commits/c5e14f5c0d30f63a9e453e3bd04e9d1795099af8 and the updates to the database.

Default values for positional arguments follow a rule, so if an argument has a default value, later positional arguments must also have default values.

The database only stores the actual default values, and nothing about the arguments that doesn't have default values.

This turns out to be a major problem for `Argument.getKwDefault(i)`, since default values for keyword-only arguments doesn't have the same rule. So if you know there is one default value, you can't tell if it is associated with `foo` or `bar`, as in the examples below:

```py
def a(*, foo=None, bar):
    pass

def b(*, foo, bar=None):
    pass
```

This PR (and corresponding internal extractor change) fixes this, by making `Argument.getKwDefault(i)` return the default value for keyword-only parameter `i` (instead of the i'th default for a keyword-only parameter). `Argument.getDefault` is changed in the same manner to keep consistency.

### CODE REVIEW

**Comment by RasmusWL** [comment]
I tested that the upgrade script works properly, 

- first off by looking at `python/ql/test/3/library-tests/functions/test.py`
- I also looked at the CodeQL DB for `django` and checked 
    - the number of `Arguments.getAKwDefault` did not change from original DB when running the query from master-branch, to the upgraded db (643 results). 
    - I also checked that with the new library functions in this PR,  `Function.getKeywordOnlyArg(_).getDefault()` gave the same number of results.

**Comment by RasmusWL** [comment]
Woops, seems like I forgot the part about _2. Run a [Stats/CPP](https://jenkins.internal.semmle.com/job/Stats/job/CPP/build) job and update ql repo with the results._ :thinking: will have to figure out if it is required for this change or not :thinking:

**Comment by RasmusWL** [comment]
Our [internal docs](https://wiki.semmle.com/pages/viewpage.action?spaceKey=IN&title=Stats+Files) say that:

> QL queries are compiled in the context of a database schema and a database statistics file. The former provides definitions of types and existential relations, while the latter provides information to the optimiser. Note in particular that QL queries are not compiled in the context of a populated database; the role of the statistics file is to be a distilled version of a prototypical database. If a statistics file is not provided, is inaccurate, or is incomplete, then the optimiser can transform queries so that they perform extremely badly.

Since this upgrade only renumbers a few items in the `py_exprs` relation, the cardinality clearly doesn't change. The histograms (relating how much the values in one column depend upon the values in the other), shouldn't change in any meaningful way either.

I think it's probably worth it to look into whether we should update the `.dbscheme.stats` file, but if we can keep it out of this PR that would be great :+1:

File: python/ql/src/semmle/python/AstGenerated.qll [hunk 1]
```diff
@@ -1191,13 +1191,13 @@ library class AliasList_ extends @py_alias_list {
 }
 
 library class Arguments_ extends @py_arguments {
-    /** Gets the keyword default values of this parameters definition. */
+    /** Gets the keyword-only default values of this parameters definition. */
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/src/semmle/python/AstGenerated.qll:1194]: Well I made the changes in the file we use to autogenerate from ... see #36846 on the internal repo

File: python/ql/src/semmle/python/Function.qll [hunk 2]
```diff
@@ -90,7 +94,7 @@ class Function extends Function_, Scope, AstNode {
     int getPositionalParameterCount() { result = count(this.getAnArg()) }
 
     /** Gets the number of keyword-only parameters */
-    int getKeywordOnlyParameterCount() { result = count(this.getAKwonlyarg()) }
+    int getKeywordOnlyParameterCount() { result = count(this.getAKeywordOnlyArg()) }
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/ql/src/semmle/python/Function.qll:97]: Well, as I detailed in the [commit message](https://github.com/github/codeql/pull/3362/commits/8c1cfe52f6cf53f9672225c34f9b317e9bf74eb8), this is purely out for consistency, so we only use the `Kw` internally

File: python/upgrades/f635b392038a494915307f913657cd3058f9b476/py_exprs.ql [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+        )
+        or
+        // expr is a default for a keyword-only parameter
+        exists(Arguments_ args, CallableExprAdjusted callable |
+            callable.getArgs() = args and
+            args.getKwDefault(oldidx) = expr and
+            newidx =
+                max(int i |
+                    exists(Parameter_ param | param = callable.getInnerScope().getKwonlyarg(i) |
+                        param.getLocation().getStartLine() < expr.getLocation().getStartLine()
```
[the remark this answers is not in the document]
Comment (RasmusWL) [reply python/upgrades/f635b392038a494915307f913657cd3058f9b476/py_exprs.ql:153]: magic

But I tried to explain in 5a18b08 :)

File: python/ql/src/semmle/python/AstGenerated.qll (hunk 1 above)
[the remark this answers is not in the document]
Comment (tausbn) [reply python/ql/src/semmle/python/AstGenerated.qll:1194]: Ah, I had completely forgotten about that! Alright, no problems here, then. :slightly_smiling_face:

**Pull request merged**